src/app/dashboard/dashboard.component.ts

```diff
@@ -1,23 +1,106 @@
-import { Component, OnInit } from '@angular/core';
-import { Hero } from '../hero';
-import { HeroService } from '../hero.service';
+import { takeUntil } from 'rxjs/operators';
+import { Component, OnDestroy, OnInit } from '@angular/core';
+import { Subject } from 'rxjs';
+import { Weapon } from '../core/interface/weapon';
+import { cloneData } from '../core/utils/utils';
+import { Hero } from '../core/interface/hero';
+import { WeaponService } from '../weapon/weapon.service';
+import { GameService } from './game.service';
+import { MessageService } from '../message.service';
+import { HeroService } from '../heroes/hero.service';
 
 @Component({
   selector: 'app-dashboard',
   templateUrl: './dashboard.component.html',
   styleUrls: [ './dashboard.component.css' ]
 })
-export class DashboardComponent implements OnInit {
+export class DashboardComponent implements OnInit, OnDestroy {
   heroes: Hero[] = [];
+  weapons: Weapon[] = [];
+  heroesPlaying: Hero[] = [];
+  $unSubscribe = new Subject();
 
-  constructor(private heroService: HeroService) { }
+  constructor(
+    private readonly heroService: HeroService,
+    private readonly weaponService: WeaponService,
+    private readonly gameService: GameService,
+    private readonly messageService: MessageService
+  ) { }
 
   ngOnInit() {
+    this.heroesPlaying = this.gameService.heroesPlaying.value || [];
     this.getHeroes();
+    this.getWeapons();
+    this.gameService.heroesPlaying
+    .pipe(takeUntil(this.$unSubscribe))
+    .subscribe((heroes: Hero[]) => {
+      this.heroesPlaying = heroes;
+    });
   }
 
+  ngOnDestroy(): void {
+    this.$unSubscribe.next();
+  }
+
+  /**
+   * get heroes
+   */
   getHeroes(): void {
     this.heroService.getHeroes()
-      .subscribe(heroes => this.heroes = heroes.slice(1, 5));
+      .subscribe(heroes => this.heroes = heroes);
+  }
+
+  /**
+   * get weapons
+   */
+  getWeapons(): void {
+    this.weaponService.getWeapons()
+      .subscribe(weapons => this.weapons = weapons);
```

Comment:
See https://medium.com/@paynoattn/3-common-mistakes-i-see-people-use-in-rx-and-the-observable-pattern-ba55fee3d031

Search for string "You are taking data outside of the Observable stream." to see the mistake here

Reply:
Thanks for your suggestion. I will read and improve my code.